Initial work on generators for #353 Data set simulators/generators

sjmiller8182 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the extension guidelines: https://github.com/alan-turing-institute/sktime/blob/master/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Initial work on generators based on mention in #353 with @mloning.

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented. 
-->

This introduces univariate generators for random normal noise, ARMA process, and linear process with noise.

#### Does your contribution introduce a new dependency? If yes, which one? 

<!--
If your contribution does add a new dependency, we may suggest to initially develop your contribution in a separate
 companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum. 
-->

No new dependencies.

#### PR checklist for new estimators 
<!--
This is only relevant if you contribute a new estimator
 (classifiers, regressors, forecasters, etc
.). If so, please go through the checklist below. Otherwise please ignore check list.

- [ ] I've added unit tests and made sure they pass locally. 
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.
- [ ] I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/master/CODEOWNERS) and I'm committed to maintain my algorithm.
- [ ] I've added myself to the list of contributors
-->

Not a new estimator.

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.

Thanks for contributing!
-->

mloning commented:
Thanks for the PR! 

Quick comment: Linting failed, you need to include `__all__ = [...]` in the the `__init__.py` file, otherwise flake8 will complain about unused imports.

mloning commented:
Thanks again @sjmiller8182 for the PR! Looks all good to me. 

@fkiraly any comments?

sjmiller8182 commented:
@mloning Thanks for taking a look!

I noticed a major copy and paste error in the docs - I'll fix that this weekend. 

Also, I want to push a few more items. I've come up with a way to change the RandomState sampling method and pass the kwargs to the sampling method. I'd also like to implement something for seasonal data generation. I'm hoping to have something to add this weekend on these two items.

I am not sure if the current code needs some kind of unit tests, since its rather simple. But, I can put something in if that is recommended.

mloning commented:
Unit tests would be nice to have, but since we're using statsmodels under the hood not we don't have to test the sampling itself I think. It would be nice to integrate the generators in some tutorial or user guide (see #361)

fkiraly commented:
>@fkiraly any comments?

* in terms of module structure, I would separate data sets from data generators - these are different things in my opinion, the former are fixed data sets, the latter are stochastic samplers of data sets. `sktime.datagenerators`?
* this introduces a dependency on `typing` and `abc` which we've tried to avoid so far. May be worth a discussion going forward, i.e., whether to integrate these in the general package design, but if not we may want to remove these as deps.
* I don't think 100 is a sensible default for `n_sample`, since arbitrary, should we rather not have a default?
* default for `burnin` in ARMA is 0, which is a "bad" default - typically, you would have a longer burn-in to ensure you are sampling from something closer to the stationary process
* the `scale` parameter should be supported, since this is another typical parameter for the mentioned processes
* `statsmodels` also allows generation of multiple series, through the `nsample` param which can be tuple valued. This might be useful for the time series classification context, so I wonder whether we also want to support this.

mloning commented:
Thanks @fkiraly!

* `typing` and `abc` come with the standard Python distribution and don't require us to install other packages, I also tend to view this as maintainer preference now and wouldn't mind having different conventions across modules, but happy to discuss this more generally
* the generators are implemented in their own file, we simply expose them to user through `sktime.datasets`, I think that's fine, the function/class names distinguish clearly whether it's *loading* or *generating* data, that way we'd follow sklearn too.

sjmiller8182 commented:
* Seems fine to me to drop `typing` since it was added in python 3.5. Is there an official list of supported python versions? `abc` is supported by python 3.
* I will add support for the kwargs of `np.random.normal` (explicitly mentioned method) into `NoiseGenerator`. However, would it make sense to create a more general version of the noise generator that can generate from any of the methods of `np.RandomState` with a `**kwargs`-like interface for setting the arguments?
* For `burnin`, should a user input be required or is there a recommendation for a default?
* I will extend the generators to accept tuples for `n_sample` and fine by me to remove the default value.

mloning commented:
Great! FYI We support Python 3.6-3.8, so typing and abc are fine.

fkiraly commented:
>Great! FYI We support Python 3.6-3.8, so typing and abc are fine.

Fine with me to use `typing`, given @mloning´s comment.
I´d like to have consistent use of `typing` and/or types at some point, but there´s no need to be that strict right now.

>However, would it make sense to create a more general version of the noise generator that can generate from any of the methods of np.RandomState with a **kwargs-like interface for setting the arguments?

Hm, that actually makes sense - as long as it does not lead to interface overkill (or unclear defaults).

>For burnin, should a user input be required or is there a recommendation for a default?

hm, `statsmodels` also has 0 as default for `burnin`. 
We could leave it at that - any other choice might require justification anad not generalize well to all use cases.

fkiraly reviewed (changes requested):
Given discussion above, I'll wait for the one change that is outstanding (`n_sample` support), please ask for a review then.
The other changes (removal/handling of `typing`, `burnin` default change) are not necessary for my approval since I'm happy with current state after discussion.

fkiraly commented:
@sjmiller8182, we are discussing general design of data set generators in #759 - are you still working on this?

There was just a minor change outstanding here.

sjmiller8182 commented:
@fkiraly, Sorry for the delay on this. If this still fits in the general scope of data generators, I will work on getting it finished.

fkiraly commented:
no worries, of course!

You may also want to chip in to the design discussion and/or ensure API consistency on your PR.

sjmiller8182 commented:
@fkiraly, @mloning,

I've implemented allowing a tuple input to `n_sample` for `NoiseGenerator` and `ArmaGenerator`, but I am not sure of a good way to implement the change for `LinearGenerator`. I may just not be looking at it correctly. I also removed the default value for `n_sample` since we had discussed previously that it doesn't make sense. 

Right now I am seeing a failure on the macos-tests, but it doesn't seem to be related to my additions so I am not sure what is causing the problem on those builds at the moment. I would appreciate it if someone could help me understand what is failing on those builds.

I also still need to look at how the unit tests are implemented here. I am not familiar with the testing structure in this project so it may take me some time to implement the tests. However, the doc tests are passing.

fkiraly commented:
Can you explain what your speicifc question about `LinearGenerator` is?

Questions from my side (not change requests):
* should the `sample` method not have two arguments, `n_sample` and `n_instance`? One for the length of the time series, one for how many there are.
* should the base class inherit from `sklearn's` `BaseEstimator`? That way, you get parameter handling for free.

Also, I strongly suggest adhering to a joint interface specificationm i.e., type and arguments of `sample` should be the same across the generators as long as it generates the same data object(s). For example, `burnin` in `sample` of `ARMAGenerator` should move to `__init__`; the types of all `sample` methods should be the same, why is this different between `ARMAGenereator` and `LinearGenerator`?

sjmiller8182 commented:
Ah yes, agreed the `sample` method should be the same between the classes. 

Actually, if we change `sample` to take two arguments, `n_sample` and `n_instance` that eliminates my issue with the `LinearGenerator`. I'll take a in cooperating `BaseEstimator` from `sklearn`.

mloning commented on sktime/datasets/generators.py:
```diff
@@ -12,21 +12,37 @@
 
     @abstractmethod
-    def sample(self, n_sample):
+    def sample(self, n_sample, n_instance):
```
```suggestion
    def sample(self, n_instances=None, n_variables=None, n_timepoints=None):
```
with some useful default values

Alternatively we could distinguish between generators for different kinds of data (time series and panel data)?

mloning reviewed (commented):
Thanks @sjmiller8182 - we also need to write some unit tests to make sure they work as expected.

sjmiller8182 commented:
@mloning, 

I've added tests. 

For my understanding, by time series you mean a single instance of a univariate or multivariate time series and by panel data you mean multiple instances of a time series? If that is the case, I think we could add a method `sample_panel` to generate the panels.

TonyBagnall commented:
hi, @sjmiller8182 we are having a tidy up of pull requests. Are you still working on this? If not, is it ok to close it? We would welcome a new pull request if you update it later

sjmiller8182 commented:
@TonyBagnall, I have implemented the requests mentioned above. I believe it is ready to be integrated.

TonyBagnall commented:
hi @sjmiller8182 sorry for the delay, if you want it to go in could you change the status so it is no longer a draft, we will then review it
